runtime/executor/src/main/java/org/apache/nemo/runtime/executor/data/partition/SerializedPartition.java

```diff
@@ -144,18 +171,41 @@ public boolean isSerialized() {
   }
 
   /**
+   * This method should only be used when this partition is constructed with a data from existing file.
+   *
    * @return the serialized data.
    * @throws IOException if the partition is not committed yet.
    */
   @Override
   public byte[] getData() throws IOException {
     if (!committed) {
       throw new IOException("The partition is not committed yet!");
+    } else if (offheap) {
+      throw new RuntimeException("This partition does not have on-heap data");
     } else {
       return serializedData;
     }
   }
 
+  /**
+   * This method is used to emit the output as {@link SerializedPartition}.
+   *
+   * @return the serialized data in list of {@link ByteBuffer}s
+   * @throws IOException if the partition is not committed yet.
+   */
+  public List<ByteBuffer> getBuffer() throws IOException {
+    if (!committed) {
+      throw new IOException("The partition is not committed yet!");
+    } else {
+      List<ByteBuffer> result = new ArrayList<>(dataList.size());
+      for (final ByteBuffer buffer : dataList) {
+        final ByteBuffer dupBuffer = buffer.duplicate();
+        result.add(dupBuffer);
```

Comment:
Why return a duplicate?

Reply:
Same with the comment above. There is re-reading of the data so sharing the position and limit value would cause erroneous reading.

Reply:
I see, thanks!